test/negative_tests.cpp

```diff
@@ -0,0 +1,259 @@
+/*
+ * The Following Program Consists of a bunch of Negative Tests
+ * 1. Dupilcate Bucket Put Test
+ * 2. Deleting Non Existent Bucket
+ * 3. Deleting Non Exiting  Object
+ * 4. Creation of Bucket with Invalid Name
+ * 5. Duplicate Object Creation
+ * 6. Bulk Put With Empty Object List
+*/
+#include <stdbool.h>
+#include <stdio.h>
+#include "ds3.h"
+#include "test.h"
+#include <boost/test/unit_test.hpp>
+
+
+
+
+//Testing a Duplicate Bucket Put
+
+BOOST_AUTO_TEST_CASE( put_duplicate_bucket) {
+	printf("-----Testing Duplicate Bucket Creation-------\n");
+    ds3_client* client = get_client();
+    uint64_t i;
+    bool found = false;
+    const char* bucket_name = "duplicatename_test_bucket";
+    ds3_request* request = ds3_init_put_bucket(bucket_name);
+    ds3_get_service_response* response;
+
+    ds3_error* error = ds3_put_bucket(client, request);
+
+    BOOST_CHECK(error == NULL);
+
+    ds3_free_request(request);
+    ds3_free_error(error);
+    request = ds3_init_get_service();
+
+    error = ds3_get_service(client, request, &response);
+
+    BOOST_CHECK(error == NULL);
+
+    for (i = 0; i < response->num_buckets; i++) {
+        //fprintf(stderr, "Expected Name (%s) actual (%s)\n", bucket_name, response->buckets[i].name->value);
+        if (strcmp(bucket_name, response->buckets[i].name->value) == 0) {
+            found = true;
+            break;
+        }
+    }
+
+    ds3_free_request(request);
+    ds3_free_error(error);
+    ds3_free_service_response(response);
+
+    BOOST_CHECK(found);
+    
+    //Putting Duplicate Bucket
+    request = ds3_init_put_bucket(bucket_name);
+    error = ds3_put_bucket(client, request);
+    if (error != NULL) {
+		printf("ds3_error status code: %lu\n", error->error->status_code);
+		printf("ds3_error: %s\n", error->error->status_message->value);
+		BOOST_CHECK(error->error->status_code == 409);
+        BOOST_CHECK(strcmp(error->error->status_message->value ,"Conflict")==0);
+        ds3_free_error(error);
+		ds3_free_request(request);
+    }
+	
+	//Deleting Created Bucket
+    request = ds3_init_delete_bucket(bucket_name);
+    error = ds3_delete_bucket(client, request);
+    ds3_free_request(request);
+
+    free_client(client);
+    BOOST_CHECK(error == NULL);
```

Comment:
I would replace `BOOST_CHECK(error == NULL)` with `handle_error(error)`  The handle error call will print out some information about the error and then free it. By extension this makes the following `ds3_free_error` call unneeded.